source/plugins/go/src/oms.go

```diff
@@ -746,70 +860,144 @@ func PostTelegrafMetricsToLA(telegrafRecords []map[interface{}]interface{}) int
 		Log(message)
 	}
+	
+	if IsWindows == false { //for linux, mdsd route
+		var msgPackEntries []MsgPackEntry			
+		var i int
+		start := time.Now()
+        var elapsed time.Duration
+
+		for i = 0; i < len(laMetrics); i++ { 
+				var interfaceMap map[string]interface{} 
+				stringMap := make(map[string]string)
+				jsonBytes, err := json.Marshal(*laMetrics[i])
+				if err != nil {
+					message := fmt.Sprintf("PostTelegrafMetricsToLA::Error:when marshalling json %q", err)
+					Log(message)
+					SendException(message)
+					return output.FLB_OK
+				} else {
+					if err := json.Unmarshal(jsonBytes, &interfaceMap); err != nil { 							
+						message := fmt.Sprintf("Error while UnMarshalling json bytes to interfaceMap: %s", err.Error())
+						Log(message)
+						SendException(message)
+						return output.FLB_OK
+					} else {	
+						for key, value := range interfaceMap {
+							strKey := fmt.Sprintf("%v", key)
+							strValue := fmt.Sprintf("%v", value)
+							stringMap[strKey] = strValue
+						}				
+						msgPackEntry := MsgPackEntry{							
+							Record: stringMap,
+						}
+				     	msgPackEntries = append(msgPackEntries, msgPackEntry) 
+					}				
+				}
+		}
+		if (len(msgPackEntries) > 0) {							
+				msgpBytes := convertMsgPackEntriesToMsgpBytes(MdsdInsightsMetricsTagName, msgPackEntries)			
+				if MdsdInsightsMetricsMsgpUnixSocketClient == nil {
+					Log("Error::mdsd::mdsd connection does not exist. re-connecting ...")
+					CreateMDSDClientInsightsMetrics(ContainerType)
+					if MdsdInsightsMetricsMsgpUnixSocketClient == nil {
+						Log("Error::mdsd::Unable to create mdsd client for insights metrics. Please check error log.")					
+						ContainerLogTelemetryMutex.Lock()
+						defer ContainerLogTelemetryMutex.Unlock()
+						InsightsMetricsMDSDClientCreateErrors += 1
+						return output.FLB_RETRY
+					}
+				}
 
-	var metrics []laTelegrafMetric
-	var i int
+				deadline := 10 * time.Second
+				MdsdInsightsMetricsMsgpUnixSocketClient.SetWriteDeadline(time.Now().Add(deadline)) //this is based of clock time, so cannot reuse			
+				bts, er := MdsdInsightsMetricsMsgpUnixSocketClient.Write(msgpBytes)
 
-	for i = 0; i < len(laMetrics); i++ {
-		metrics = append(metrics, *laMetrics[i])
-	}
+				elapsed = time.Since(start)
 
-	laTelegrafMetrics := InsightsMetricsBlob{
-		DataType:  InsightsMetricsDataType,
-		IPName:    IPName,
-		DataItems: metrics}
+				if er != nil {
+					Log("Error::mdsd::Failed to write to mdsd %d records after %s. Will retry ... error : %s", len(msgPackEntries), elapsed, er.Error())
+					if MdsdInsightsMetricsMsgpUnixSocketClient != nil {
+						MdsdInsightsMetricsMsgpUnixSocketClient.Close()
+						MdsdInsightsMetricsMsgpUnixSocketClient = nil
+					}
 
-	jsonBytes, err := json.Marshal(laTelegrafMetrics)
+					ContainerLogTelemetryMutex.Lock()
+					defer ContainerLogTelemetryMutex.Unlock()
```

Comment:
Can you look into all mutex-es ? I see single mutex used in multiple streams and they are all sourced by different plugins/inputs and each stream will wait on other streams to aquire the lock.

Reply:
reviewed & looks good to me. we already use the ContainerLogTelemetryMutex for both Telegraf & containerlog related to telemetry. using the same pattern. let me know if I am missing anything which could potential cause deadlock.